source/agora/consensus/Validation.d

```diff
@@ -631,6 +631,268 @@ unittest
     }
 }
 
+/// Test for transactions having no input or no output
+unittest
+{
+    import std.format;
+    Transaction[Hash] storage;
+    KeyPair key_pair = KeyPair.random;
+
+    // delegate for finding `Output`
+    scope findUTXO = (Hash hash, size_t index, out UTXOSetValue value)
+    {
+        if (auto tx = hash in storage)
+        {
+            if (index < tx.outputs.length)
+            {
+                value.unlock_height = 0;
+                value.type = TxType.Payment;
+                value.output = tx.outputs[index];
+                return true;
+            }
+        }
+        return false;
+    };
+
+    // Creates the transaction having no input
+    Transaction oneTx = Transaction(
+        TxType.Payment,
+        [],
+        [Output(Amount(50), key_pair.address)]
+    );
+    Hash oneHash = hashFull(oneTx);
+    storage[oneHash] = oneTx;
+
+    // Test for Payment transaction having no input
+    assert(!oneTx.isValid(findUTXO, 0), 
+        format("Tx having no input is not validated %s", oneTx));
```

Comment:
A general rule on assert message: Consider how it's going to show up in the console.
In this case, another developer would see: "Assertion failure: Tx having no input is not validated".
This is quite ambiguous, as if the message describes what should happen, not what is happening.
Consider the following wording:
- "Tx having no input should not pass validation"
- "Tx having no input are considered valid"

Those two wordings might not be perfect, but are much less ambiguous: the first one define what should happen, and a reader will understand that it is *not* happening (because the assertion failed), or what is happening, and hopefully that will be clear enough that the reader can infer what is wrong.